cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatusTool.java

```diff
@@ -378,13 +442,85 @@ private static void dumpReplicationStateTable() throws Exception {
     }
   }
 
-  private static void dumpMap(HashMap<String, Long> regionMap) {
-    for (Map.Entry<String, Long> entry: regionMap.entrySet()) {
-      System.out.println(entry.getKey() + "=>" + " timestamp=" + entry.getValue());
+  private static void checkHDFSReplicationComplete(SortedMap<String, String> masterChecksumMap) throws IOException {
+    boolean complete;
+    SortedMap<String, String> slaveChecksumMap = getClusterChecksumMap();
+
+    //Check if all files are accounted on both clusters
+    if (masterChecksumMap.size() != slaveChecksumMap.size()) {
+      System.out.println("Number of HDFS files on the Master and Slave Clusters do not match.");
+    }
+
+    complete = !checkDifferences(masterChecksumMap.keySet(), slaveChecksumMap.keySet(), "File");
+
+    for (Map.Entry<String, String> checksumEntry : masterChecksumMap.entrySet()) {
+      String file = checksumEntry.getKey();
+      String masterChecksum = checksumEntry.getValue();
+      String slaveChecksum = slaveChecksumMap.get(file);
+
+      if (slaveChecksum != null && !masterChecksum.equals(slaveChecksum)) {
+        System.out.println("Master Checksum " + masterChecksum + " for File " + file + " does not match with"
+                             + " Slave Checksum " + slaveChecksum);
+        complete = false;
+      }
     }
+
+    if (complete) {
+      // If checksums match for all files.
+      System.out.println("Master and Slave Checksums match. HDFS Replication is complete.");
+    } else {
+      System.out.println("HDFS Replication is NOT Complete.");
+    }
+  }
+
+  private static String normalizedFileName(String fileName) {
+    //while matching the filenames, ignore the IP addresses of the clusters
+    String normalizedName;
+    String ipv4Pattern = "(([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.){3}([01]?\\d\\d?|2[0-4]\\d|25[0-5])";
+    String ipv6Pattern = "([0-9a-f]{1,4}:){7}([0-9a-f]){1,4}";
+    normalizedName = fileName.replaceFirst(ipv4Pattern, "<hostname>");
+    normalizedName = normalizedName.replaceFirst(ipv6Pattern, "<hostname>");
+    return normalizedName;
   }
 
-  private static String getClusterChecksum() throws IOException {
-    return "checksum";
+  private static SortedMap<String, String> getClusterChecksumMap() throws IOException {
```

Comment:
The way the tool is designed is to be very generic where user can provide a list of hdfs paths that need to be validated and the tool will generate checksums for all files under those directories. By default, if no input file is provided it only looks under /cdap/tx.snapshot directory. Adding checks for snapshot files breaks this generality. The admin is expected to copy all hdfs files under these directories and the files don't change after cdap has been shut down [which is the first test for the tool]
Although, the last snapshot and all the log files after the last snapshot are sufficient for testing snapshot correctness, the code covers all files in favor of generality of code and scalability of the tool.